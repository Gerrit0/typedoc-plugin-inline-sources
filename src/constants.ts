export const TAG_NAME = '@source';
