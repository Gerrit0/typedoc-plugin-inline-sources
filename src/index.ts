import {
  Application,
  CommentTag,
  Context,
  Converter,
  Reflection,
} from 'typedoc';
import * as ts from "typescript"

export function load(app: Application) {
  app.converter.on(
    Converter.EVENT_CREATE_DECLARATION,
    (context: Context, reflection: Reflection) => {
      injectSource(context, reflection);
    }
  );
  app.converter.on(
    Converter.EVENT_CREATE_SIGNATURE,
    (context: Context, reflection: Reflection) => {
      injectSource(context, reflection);
    }
  );
}

function injectSource(context: Context, reflection: Reflection) {
  // ignore project declaration
  if(reflection.isProject()) return

  // get code from the reflection
  const sym = reflection.variant == "signature" ? context.project.getSymbolFromReflection(reflection.parent!) : context.project.getSymbolFromReflection(reflection)
  if(!sym) return

  const valueDeclaration = (sym.declarations || [])[0]
  if(!valueDeclaration) return

  let code: string
  if(valueDeclaration.kind == ts.SyntaxKind.VariableDeclaration) {
    // the VariableDeclaration does not include the `const` keyword because this belongs to
    // the VariableDeclarationList
    code = `const${valueDeclaration.getFullText()}`
  } else {
    code = valueDeclaration.getFullText()
  }

  if (reflection.comment?.blockTags) {
    reflection.comment.blockTags = reflection.comment.blockTags.map((tag) => {
      if (tag.tag === '@source') {
        return mapSourceTag(tag, code);
      } else {
        return tag;
      }
    });
  }
}

function mapSourceTag(
  tag: CommentTag,
  code: string
) {

  return new CommentTag("@source", [
    {
      kind: "code",
      text: `${tag.content[0]?.text || ''}\n\`\`\`typescript\n${stripBlockComments(
        code
      )}\n\`\`\`\n\n`
    }
  ])
}

function stripBlockComments(code: string) {
  return code
    .replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/, '$1')
    .replace(/^\s+|\s+$/g, '');
}
