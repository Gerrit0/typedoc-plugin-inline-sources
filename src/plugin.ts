import { Context, Reflection } from 'typedoc';
import ts from 'typescript';
import { getValueDeclaration, parseTag } from './helpers.js';

export default (
  context: Context,
  reflection: Reflection,
  sig?: ts.SignatureDeclaration,
) => {
  const TAG_NAME = '@source';

  // ignore projects
  if (reflection.isProject()) return;

  // get value declaration
  const valueDeclaration = sig || getValueDeclaration(context, reflection);

  if (!valueDeclaration) return;

  // map comment blocks
  if (reflection.comment?.blockTags) {
    reflection.comment.blockTags = reflection.comment.blockTags.map((tag) => {
      if (tag.tag === TAG_NAME) {
        return parseTag(tag, sig || valueDeclaration);
      } else {
        return tag;
      }
    });
  }
};
