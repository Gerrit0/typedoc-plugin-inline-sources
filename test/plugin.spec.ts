import { Application, DeclarationReflection, ProjectReflection } from 'typedoc';
import { load } from "../src"

describe(`Plugin:`, () => {
  let app: Application;
  let project: ProjectReflection;

  beforeAll(async () => {
    app = await Application.bootstrap({
      tsconfig: './tsconfig.test.json',
    })
    load(app)
    project = await app.convert() as ProjectReflection;
  });

  test(`should inject variable source`, async () => {
    const reflection = project.getChildByName("message")
    expect(reflection?.comment?.blockTags).toMatchSnapshot();
  });

  test(`should inject type alias source`, async () => {
    const reflection = project.getChildByName("ID")
    expect(reflection?.comment?.blockTags).toMatchSnapshot();
  });

  test(`should inject function source`, async () => {
    const reflection = project.getChildByName("doSomething") as DeclarationReflection
    expect(reflection.signatures![0].comment?.blockTags).toMatchSnapshot();
  });

  test(`should inject class source`, async () => {
    const reflection = project.getChildByName("GoodGreeter")
    expect(reflection?.comment?.blockTags).toMatchSnapshot();
  });

  test(`should inject class method source`, async () => {
    const reflection = project.getChildByName("Rectangle")?.getChildByName("calcArea") as DeclarationReflection
    expect(reflection.signatures![0].comment?.blockTags).toMatchSnapshot();
  });

  test(`should inject interface source`, async () => {
    const reflection = project?.getChildByName('Person');
    expect(reflection?.comment?.blockTags).toMatchSnapshot();
  });
});
