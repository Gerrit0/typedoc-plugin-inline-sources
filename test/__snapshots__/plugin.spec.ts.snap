// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin: should inject class method source 1`] = `
Array [
  Object {
    "paramName": "",
    "tagName": "source",
    "text": "Class method source

\`\`\`typescript
calcArea() {
    return this.height * this.width;
}
\`\`\`

",
  },
]
`;

exports[`Plugin: should inject class source 1`] = `
[
  CommentTag {
    "content": [
      {
        "kind": "code",
        "text": "Class source
\`\`\`typescript
export class GoodGreeter {
  name: string;

  constructor() {
    this.name = 'hello';
  }
}
\`\`\`

",
      },
    ],
    "tag": "@source",
  },
]
`;

exports[`Plugin: should inject function source 1`] = `
Array [
  Object {
    "paramName": "",
    "tagName": "source",
    "text": "Function source

\`\`\`typescript
export function doSomething(x: string | null) {
    if (x === null) {
        /* do nothing */
    }
    else {
        // do something
        console.log('Hello, ' + x.toUpperCase());
    }
}
\`\`\`

",
  },
]
`;

exports[`Plugin: should inject interface source 1`] = `
[
  CommentTag {
    "content": [
      {
        "kind": "code",
        "text": "Interface source
\`\`\`typescript
export interface Person {
  name: string;
  age: number;
}
\`\`\`

",
      },
    ],
    "tag": "@source",
  },
]
`;

exports[`Plugin: should inject type alias source 1`] = `
[
  CommentTag {
    "content": [
      {
        "kind": "code",
        "text": "Type Alias source
\`\`\`typescript
export type ID = number | string;
\`\`\`

",
      },
    ],
    "tag": "@source",
  },
]
`;

exports[`Plugin: should inject variable source 1`] = `
[
  CommentTag {
    "content": [
      {
        "kind": "code",
        "text": "source tag comment?
\`\`\`typescript
const message = 'Hello World!'
\`\`\`

",
      },
    ],
    "tag": "@source",
  },
]
`;
